File: app/src/main/java/com/raju/disney/opentelemetry/CrashReporter.java
package com.raju.disney.opentelemetry;

import androidx.annotation.NonNull;

import java.io.PrintWriter;
import java.io.StringWriter;

import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.StatusCode;
import io.opentelemetry.api.trace.Tracer;
import io.opentelemetry.sdk.OpenTelemetrySdk;
import io.opentelemetry.sdk.trace.SdkTracerProvider;
import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;

class CrashReporter {

    static void initializeCrashReporting(Tracer tracer, OpenTelemetrySdk openTelemetrySdk) {
        Thread.UncaughtExceptionHandler existingHandler = Thread.getDefaultUncaughtExceptionHandler();
        Thread.setDefaultUncaughtExceptionHandler(new CrashReportingExceptionHandler(tracer, openTelemetrySdk.getSdkTracerProvider(), existingHandler));
    }

    static class CrashReportingExceptionHandler implements Thread.UncaughtExceptionHandler {
        private final Tracer tracer;
        private final Thread.UncaughtExceptionHandler existingHandler;
        private final SdkTracerProvider sdkTracerProvider;

        CrashReportingExceptionHandler(Tracer tracer, SdkTracerProvider sdkTracerProvider, Thread.UncaughtExceptionHandler existingHandler) {
            this.tracer = tracer;
            this.existingHandler = existingHandler;
            this.sdkTracerProvider = sdkTracerProvider;
        }

        @Override
        public void uncaughtException(@NonNull Thread t, @NonNull Throwable e) {
            StringWriter writer = new StringWriter();
            e.printStackTrace(new PrintWriter(writer));
            String exceptionType = e.getClass().getSimpleName();
            Span span = tracer.spanBuilder(exceptionType)
                    .setAttribute(SemanticAttributes.THREAD_ID, t.getId())
                    .setAttribute(SemanticAttributes.THREAD_NAME, t.getName())
                    .setAttribute(SemanticAttributes.EXCEPTION_STACKTRACE, writer.toString())
                    .setAttribute(SemanticAttributes.EXCEPTION_ESCAPED, true)
                    .setAttribute(DisneyOtel.COMPONENT_KEY, DisneyOtel.COMPONENT_CRASH)
                    .startSpan();
            DisneyOtel.addExceptionAttributes(span, e);
            span.setStatus(StatusCode.ERROR).end();
            //do our best to make sure the crash makes it out of the VM
            sdkTracerProvider.forceFlush();
            //preserve any existing behavior:
            if (existingHandler != null) {
                existingHandler.uncaughtException(t, e);
            }
        }
    }
}

File: app/src/main/java/com/raju/disney/opentelemetry/CrashReporter.kt
package com.raju.disney.opentelemetry

import com.raju.disney.opentelemetry.DisneyOTel.Companion.addExceptionAttributes
import io.opentelemetry.api.trace.StatusCode
import io.opentelemetry.api.trace.Tracer
import io.opentelemetry.sdk.OpenTelemetrySdk
import io.opentelemetry.sdk.trace.SdkTracerProvider
import io.opentelemetry.semconv.trace.attributes.SemanticAttributes
import java.io.PrintWriter
import java.io.StringWriter

internal object CrashReporter {

    fun initializeCrashReporting(tracer: Tracer, openTelemetrySdk: OpenTelemetrySdk) {
        val existingHandler = Thread.getDefaultUncaughtExceptionHandler()
        Thread.setDefaultUncaughtExceptionHandler(
            CrashReportingExceptionHandler(
                tracer,
                openTelemetrySdk.sdkTracerProvider,
                existingHandler
            )
        )
    }

    internal class CrashReportingExceptionHandler(
        private val tracer: Tracer,
        private val sdkTracerProvider: SdkTracerProvider,
        private val existingHandler: Thread.UncaughtExceptionHandler?
    ) : Thread.UncaughtExceptionHandler {

        override fun uncaughtException(t: Thread, e: Throwable) {
            val writer = StringWriter()
            e.printStackTrace(PrintWriter(writer))
            val exceptionType = e.javaClass.simpleName
            val span = tracer.spanBuilder(exceptionType)
                .setAttribute(SemanticAttributes.THREAD_ID, t.id)
                .setAttribute(SemanticAttributes.THREAD_NAME, t.name)
                .setAttribute(SemanticAttributes.EXCEPTION_STACKTRACE, writer.toString())
                .setAttribute(SemanticAttributes.EXCEPTION_ESCAPED, true)
                .setAttribute(DisneyOTel.COMPONENT_KEY, DisneyOTel.COMPONENT_CRASH)
                .startSpan()
            addExceptionAttributes(span, e)
            span.setStatus(StatusCode.ERROR).end()
            //do our best to make sure the crash makes it out of the VM
            sdkTracerProvider.forceFlush()
            //preserve any existing behavior:
            existingHandler?.uncaughtException(t, e)
        }
    }
}